services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/containers_extension.py

```diff
@@ -0,0 +1,332 @@
+
+
+@containers_router.post(
+    "/containers/ports/outputs:push",
+    summary="Push output ports data",
+    response_class=Response,
+    status_code=status.HTTP_204_NO_CONTENT,
+)
+async def push_output_ports(
+    port_keys: Optional[List[str]] = None, rabbitmq: RabbitMQ = Depends(get_rabbitmq)
+) -> None:
+    port_keys = [] if port_keys is None else port_keys
+    mounted_volumes: MountedVolumes = get_mounted_volumes()
+
+    await send_message(rabbitmq, f"Pushing outputs for {port_keys}")
+    await nodeports.upload_outputs(
+        mounted_volumes.disk_outputs_path, port_keys=port_keys
+    )
+    await send_message(rabbitmq, "Finished pulling outputs")
+
+
+@containers_router.post(
+    "/containers:restart",
+    response_class=Response,
+    status_code=status.HTTP_204_NO_CONTENT,
+    responses={
+        status.HTTP_404_NOT_FOUND: {"description": "Container does not exist"},
+        status.HTTP_422_UNPROCESSABLE_ENTITY: {
+            "description": "Error while running docker-compose command"
+        },
+    },
+)
+async def restarts_containers(
+    command_timeout: float = Query(
+        10.0, description="docker-compose stop command timeout default"
+    ),
+    app: FastAPI = Depends(get_application),
+    settings: DynamicSidecarSettings = Depends(get_settings),
+    shared_store: SharedStore = Depends(get_shared_store),
+    rabbitmq: RabbitMQ = Depends(get_rabbitmq),
+) -> None:
+    """Removes the previously started service
+    and returns the docker-compose output"""
+
+    stored_compose_content = shared_store.compose_spec
+    if stored_compose_content is None:
+        raise HTTPException(
+            status.HTTP_404_NOT_FOUND,
+            detail="No spec for docker-compose command was found",
+        )
+
+    for container_name in shared_store.container_names:
+        await stop_log_fetching(app, container_name)
+
+    command = (
+        "docker-compose --project-name {project} --file {file_path} "
+        "restart --timeout {stop_and_remove_timeout}"
+    )
+
+    finished_without_errors, stdout = await write_file_and_run_command(
+        settings=settings,
+        file_content=stored_compose_content,
+        command=command,
+        command_timeout=command_timeout,
+    )
+    if not finished_without_errors:
+        error_message = (f"'{command}' finished with errors\n{stdout}",)
+        logger.warning(error_message)
+        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=stdout)
+
+    for container_name in shared_store.container_names:
+        await start_log_fetching(app, container_name)
+
+    await send_message(rabbitmq, "Service was restarted please reload the UI")
+    await rabbitmq.send_event_reload_iframe()
+
+
+@containers_router.post(
+    "/containers/{id}/networks:attach",
+    summary="attach container to a network, if not already attached",
+    response_class=Response,
+    status_code=status.HTTP_204_NO_CONTENT,
+)
+async def attach_container_to_network(
```

Comment:
were there any change in that file? why is it completely new? did this get separated?

Reply:
I had to split it for te line count